### PR TITLE
dex Operator Example

### PR DESCRIPTION
**Pull request by theishshah**
This operator is designed to run a dex service. Currently is defaults to the mock connector and the user can control the number of replicas of the dex pods. Future work involves allowing for configurable identity connectors in the CRD.

Tested by running the operator and changing the size field in the CRD to ensure that the correct number of replicas is deployed at any given time.

### CODE REVIEW

**Comment by hasbro17** [comment]
@theishshah Can you add a README for this to explain what this operator is and possibly a walkthrough of how to deploy it and use it(i.e create a CR and verify Dex is running). 
I can see its mostly just deploying Dex and then ensuring the deployment size is correct.

@rithujohn191 Can you review this as well since you're more familiar with Dex.

**Comment by rithujohn191** [comment]
@theishshah could you please update the description of this PR with the operator details

File: dex-operator/.gitignore [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+tags
+
+### VisualStudioCode ###
+.vscode/*
+!.vscode/settings.json
+!.vscode/tasks.json
+!.vscode/launch.json
+!.vscode/extensions.json
+.history
+
```
Comment (rithujohn191) [inline dex-operator/.gitignore]: Please scrape this down to the bare bones. A lot of this is Visual Studio specific and we dont want to check that into the repo.

File: dex-operator/README.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
+# Dex Operator
+
```
Comment (rithujohn191) [inline dex-operator/README.md]: Could you please add a small description to what Dex is and insert a link to the project

File: dex-operator/README.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+## Overview
+This Dex operator is an operator built with the [Operator SDK][operator_sdk]. Currently the operator supports scaling the numbers of pods on the mock connector and future work involves allowing the connectors to be configured using the operator CRD.
```
Comment (rithujohn191) [inline dex-operator/README.md]: You are essentially scaling the Dex pods not the connector. This statement is a little misleading. Please clarify that the Dex pod is essentially the server.

File: dex-operator/README.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+$ dep ensure
+```
+
+### Build the operator
+Build the Memcached operator image and push it to a public registry, such as quay.io
+```
+$ export IMAGE=quay.io/example-inc/dex-operator:v0.0.1
+$ operator-sdk build $IMAGE
+$ docker push $IMAGE
+```
```
Comment (rithujohn191) [inline dex-operator/README.md]: After you build it please add a section describing what the operator will do or a workflow of its operations.

File: dex-operator/cmd/dex-operator/main.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+func main() {
+	printVersion()
+
+	sdk.ExposeMetricsPort()
+
+	resource := "auth.example.com/v1alpha1"
+	kind := "Dex"
+	namespace, err := k8sutil.GetWatchNamespace()
+	if err != nil {
+		logrus.Fatalf("Failed to get watch namespace: %v", err)
```
Comment (rithujohn191) [inline dex-operator/cmd/dex-operator/main.go]: This doesn't seem like the right way to used structures logging. You need to first initialize a logger with a config. See: https://github.com/coreos/dex/blob/6ef8cd512fb9e075bf7d70ec3fe74a8397eadde4/cmd/dex/serve.go#L66

File: dex-operator/tmp/codegen/boilerplate.go.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+
```
[the remark this answers is not in the document]
Comment (fanminshi) [reply dex-operator/tmp/codegen/boilerplate.go.txt]: @rithujohn191 this is generated by the sdk. I believe it is used by the k8s codegen to prepend license to the generated code.

[the remark this answers is not in the document]
Comment (rithujohn191) [reply dex-operator/tmp/codegen/boilerplate.go.txt]: Oh I see. Thanks.

File: dex-operator/pkg/apis/auth/v1alpha1/types.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+type Dex struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              DexSpec   `json:"spec"`
+	Status            DexStatus `json:"status,omitempty"`
+}
+
+type DexSpec struct {
+	// Size is the size of the memcached deployment
```
Comment (estroz) [inline dex-operator/pkg/apis/auth/v1alpha1/types.go]: `s/\[Mm\]emcached/\[Dd\]ex/g`

File: dex-operator/pkg/stub/handler.go [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+		if event.Deleted {
+			return nil
+		}
+		dep := newDexPod(o)
+		err := sdk.Create(dep)
+		if err != nil && !errors.IsAlreadyExists(err) {
+			logrus.Errorf("Failed to create dex pod : %v", err)
+			return err
+		}
+		err = sdk.Create(newDexPod(o))
```
Comment (estroz) [inline dex-operator/pkg/stub/handler.go]: Done above, remove one or the other.

File: dex-operator/pkg/stub/handler.go [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+				Protocol:   corev1.ProtocolTCP,
+				Port:       5556,
+				TargetPort: intstr.FromInt(5556),
+				NodePort:   32000,
+			}},
+		},
+	}
+}
+
+func newDexConfigMap(cr *v1alpha1.Dex) *corev1.ConfigMap {
```
Comment (estroz) [inline dex-operator/pkg/stub/handler.go]: This function isn't called anywhere.

File: dex-operator/pkg/stub/handler.go [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ConfigMap",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Name,
+			Namespace: cr.Namespace,
+		},
+		Data: map[string]string{
+			"config.yaml": `|
```
Comment (estroz) [inline dex-operator/pkg/stub/handler.go]: Should be read from a file in a pre-defined location, or file location should be passed in as a parameter.

**Comment by theishshah** [comment]
Save for auto-gen issues (which I will fix once we have a patch in the sdk) this operator should be ready to go

**Comment by camilamacedo86** [comment]
Closing this one as outdated since we decided to move forward just with the Memcached ones. 
However, please feel free to re-open if you think that would be required.

**Pull request closed without merging**